Bring consistency to the time coordinate names

b-reyes opened the pull request:
In this PR I address #656, which brings more consistency to the names of the time coordinates for the `Platform` and `Environment` groups. Specifically, the following changes were made. 

* Platform 
    * Let `time1.comment = "Time coordinate corresponding to GPS location."`
        * Note: AZFP does not have a `time1` 
    * Let `time2.comment = “Time coordinate corresponding to platform sensors.”`
    * Let `time3.comment = “Time coordinate corresponding to environmental variables.”`
        * For EK80 `time3.comment = “Time coordinate corresponding to environmental variables. Note that Platform.time3 is the same as Environment.time1”`
        * Note: AZFP does not have a `time3`
    * For EK60 `ping_time` was changed to `time2` for the variables `pitch/roll/vertical_offset`. The variable `time3` was created (the values are identical to `time2`) and `ping_time` was changed to `time3` for the variable `water_level`.
    * For AZFP `ping_time` was changed to `time2` 
* Environment
    * Let `time1.comment = “Time coordinate corresponding to environmental variables.”`
        * For EK80 `time1.comment = “Time coordinate corresponding to environmental variables. Note that Platform.time3 is the same as Environment.time1”`
    * For EK60 changed `ping_time` to `time1`
    * For EK80 defined time1 as `self.parser_obj.environment["timestamp"]`, if it exists and NaT, if it doesn’t. Then replaced `ping_time` and `environment_time` with `time1` 
    * For AZFP changed `ping_time` to `time1`
* Platform/NMEA 
    * Let `Platform/NMEA.time1.comment = "Time coordinate corresponding to GPS location."`
    * Note: the values of `Platform/NMEA.time1` will be changed in a different PR
* Added `"time3": DEFAULT_TIME_ENCODING` to `​​DEFAULT_ENCODINGS`
* Rename `time1` to `ping_time` in `get_env_params` for all calibration routines and modify associated tests. Additionally, renamed `time3` to `ping_time` at the end of the function `_apply` in `calibrate_base.py`. 


IMPORTANT NOTE: Since we changed some of the time names in the `Platform` and `Environment` groups, this caused some conflicts in the calibration routines, the `compute_range` function in `EchoData`, and the visualize api function `_add_water_level`. This is because in these routines we are often adding together variables from the beam groups (which use `ping_time`) and  variables from the `Platform`/`Environment` groups (which use either `time1` or `time3`). Since `ping_time` is a required dimension for the beam groups (in the convention), I thought it best to not change its name and rather account for this difference in downstream routines.

emiliom commented on echopype/calibrate/calibrate_base.py:
```diff
@@ -194,7 +194,15 @@ def _apply(self, echodata) -> Dict[str, xr.DataArray]:
 
-        return {var: env_params[var] for var in ("temperature", "salinity", "pressure")}
+        if self.data_kind == "stationary":
+            # renaming time2 to ping_time is necessary because we are performing
+            # calculations with the beam groups that use ping_time
```
This was intended to refer to time3. Also, I think it can be confusing to keep track of whether `time#` refers to a dimension from the Platform or Environment group, when you're operating on both; I added a clarification
```suggestion
            # renaming time3 (from Platform group) to ping_time is necessary because 
            # we are performing calculations with the beam groups that use ping_time
```

emiliom commented:
A style suggestion about a pattern that occurs four times in `calibrate/calibrate_ek.py` in this PR, eg:

```python
self.env_params["sound_absorption"] = (
    self.env_params["sound_absorption"]
    if "sound_absorption" in self.env_params
    # renaming time1 to ping_time is necessary because we are performing
    # calculations with the beam groups that use ping_time
    else self.echodata.environment["absorption_indicative"].rename(
        {"time1": "ping_time"}
    )
)
```

I find that the 2-line comment in the middle of the if-else statement that's already part of an assignment makes it harder to grasp the statement. It's already a somewhat long ternary statement (a if True else b). Also, the comment occurs in two pairs, where they're fairly close to one another in each pair.

I suggest moving the comments to occur only once within each surrounding/parent else block, to the start of the else block, to read something like:
```python
    # Below, renaming time1 to ping_time is necessary because
    # we are performing calculations with the beam groups that use ping_time
```

emiliom commented on echopype/visualize/api.py:
```diff
@@ -206,6 +206,11 @@ def _add_water_level(
     if isinstance(water_level, xr.DataArray):
         check_dims = range_in_meter.dims
+
+        # rename time3 to ping_time because range_in_meter is in ping_time
```
How about moving this statement to the start of the function? That way it'll also encompass the similar renaming in line 199

emiliom commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -308,15 +312,27 @@ def set_platform(self, NMEA_only=False) -> xr.Dataset:
+                                "axis": "T",
+                                "long_name": "Timestamps for position datagrams",
+                                "standard_name": "time",
+                                "comment": "Time coordinate corresponding to platform sensors.",
```
I think this comment should be "Time coordinate corresponding to platform motion and position sensors.". See @leewujung's initial comment/description in #656.

```suggestion
                                "comment": "Time coordinate corresponding to platform motion and position sensors.",
```

emiliom commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -308,15 +312,27 @@ def set_platform(self, NMEA_only=False) -> xr.Dataset:
                             self.parser_obj.ping_time[ch],
                             {
                                 "axis": "T",
                                 "long_name": "Timestamps for position datagrams",
```
`long_name` may need to be changed, otherwise it's identical (indistinguishable) to the one in `time2`. Maybe "Timestamps for environmental variables"? Though would change the nature of the `long_name` from the data origin (datagram type) to, umm, its scope.

emiliom commented on echopype/convert/set_groups_azfp.py:
```diff
@@ -120,10 +121,13 @@ def set_platform(self) -> xr.Dataset:
+                    time2,
+                    {
+                        **self._varattrs["beam_coord_default"]["ping_time"],
+                        "comment": "Time coordinate corresponding to platform sensors.",
```
See my comment for ek60

emiliom commented on echopype/convert/set_groups_ek80.py:
```diff
@@ -65,20 +65,18 @@ def __init__(self, *args, **kwargs):
 
     def set_env(self, env_only=False) -> xr.Dataset:
```
`env_only` is apparently no longer used, so it should be removed.

emiliom commented on echopype/convert/set_groups_ek80.py:
```diff
@@ -388,6 +377,7 @@ def set_platform(self) -> xr.Dataset:
                         "axis": "T",
                         "long_name": "Timestamps for NMEA datagrams",
                         "standard_name": "time",
+                        "comment": "Time coordinate corresponding to GPS location.",
```
Replace these 3 lines with what's used in ek60:

```python
                    {
                        **self._varattrs["platform_coord_default"]["time1"],
                        "comment": "Time coordinate corresponding to GPS location.",
                    },
```

The result is the same, but the code ends up being more consistent and leveraging 1.0.yml

emiliom commented on echopype/convert/set_groups_ek80.py:
```diff
@@ -366,6 +354,7 @@ def set_platform(self) -> xr.Dataset:
                         "axis": "T",
                         "long_name": "Timestamps for MRU datagrams",
                         "standard_name": "time",
+                        "comment": "Time coordinate corresponding to platform sensors.",
```
See time2 comment in ek60

emiliom commented:
I'm done reviewing. I don't think I found any substantial issue. Most of my comments are suggestions or small cleanups. Nice job.

Many, if not most of my comments are actually about the strings used in variable attributes. I'm leaning towards suggesting that for cases where we don't have a clear, easy agreement, let's postpone the resolution to a follow up PR focused on attributes only. Merging this PR soon will be very important, in order to have in place the actual, "structural" changes (to variable names) in place. `long_name` and `comment` variable attributes don't have direct impacts on downstream code (they're intended for eyeballs), so we addressing them in a follow up PR makes sense to me.

b-reyes commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -350,7 +366,7 @@ def set_platform(self, NMEA_only=False) -> xr.Dataset:
             ds = xr.merge([ds, ds_plat], combine_attrs="override")
 
-            ds = ds.chunk({"ping_time": DEFAULT_CHUNK_SIZE["ping_time"]})
+            ds = ds.chunk({"time2": DEFAULT_CHUNK_SIZE["ping_time"]})
```
> We could also just leave the chunking issue to 0.6.1. It's overdue for more attention.

Yes, I agree that this is not something we should address here. Plus, chunking both `time2` and `time3` may be inefficient for computations. Let's leave this for 0.6.1.

b-reyes commented:
@emiliom I believe I have addressed all of your concerns except the discussions on changing the `long_name` and `comment` variable attributes. I agree with your statement above that these changes can be discussed in a different issue/PR. My vote is to handle this in a different issue as this is somewhat outside the scope of this PR and the issue it addressed. If you agree, could you please create a new issue stating all of your suggested changes?

emiliom commented:
Thank you. It looks good. I'll create a new issue as you suggested. I'll get to it late this morning and will then merge the PR.

emiliom commented:
I've created the new issue #682. While I haven't copied and pasted there my individual comments (I'll do that later today), that's not a reason to delay merging this PR. So, I'll merge now.
